Fix for activating various UI items via right click

Comandeer opened the pull request:
## What is the purpose of this pull request?

Bug fix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What is the proposed changelog entry for this pull request?

```
Fixed Issues:

* [IE, Edge] Fixed: Various editor's UI elements react to right mouse button click:
	* [#2845](https://github.com/ckeditor/ckeditor-dev/issues/2845): [Rich Combo](https://ckeditor.com/cke4/addon/richcombo).
	* [#2857](https://github.com/ckeditor/ckeditor-dev/issues/2857): [List Block](https://ckeditor.com/cke4/addon/listblock).
	* [#2858](https://github.com/ckeditor/ckeditor-dev/issues/2858): [Menu](https://ckeditor.com/cke4/addon/menu).

API Changes:

* [#2845](https://github.com/ckeditor/ckeditor-dev/issues/2845): The [`CKEDITOR.tools.normalizeMouseButton`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_tools.html#method-normalizeMouseButton) was added.
* [#2975](https://github.com/ckeditor/ckeditor-dev/issues/2975): The [`CKEDITOR.dom.element#fireEventHandler`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_dom_element.html#method-fireEventHandler) was added.

```

## What changes did you make?

I've fixed ability to use rich combos, listblocks and menus with right mouse button.

To do so I had to add `CKEDITOR.tools.fireElementEventHandler` to be able to emulate `mouseup` events with correct mouse button info in keystrokes handler in `panel`.

Closes #2845.
Closes #2857.
Closes #2858.
Closes #2975.

jacekbogdanski commented on plugins/panel/plugin.js:
```diff
@@ -443,8 +443,14 @@
+							// We must pass info about clicked button (#2857).
+							CKEDITOR.tools.fireElementEventHandler( focusable, 'on' + keyAction, {
+								button: ( CKEDITOR.env.ie && ( CKEDITOR.env.version < 9 || CKEDITOR.env.ie6Compat ) ) ?
+									1 : CKEDITOR.MOUSE_BUTTON_LEFT
```
This IE8 magic number which additionally can be mistaken with a middle button click (thus `1 == CKEDITOR.MOUSE_BUTTON_MIDDLE` should be replaced with correct mapping.

Comandeer commented on tests/plugins/menu/manual/rightclick.html:
```diff
@@ -0,0 +1,12 @@
+</div>
+
+<script>
+	if ( !CKEDITOR.env.ie ) {
```
This test does not have any sense outside of IE, because it's not possible to activate button by right mouse button. In the IE it's possible only because `mouseup` is used there instead of `click`.

Comandeer replied:
TBH I'm quite sure that this kind of breakage can't be missed even without tests, as it would make any button unclickable. But added manual test for `panel` should also cover such cases.

Comandeer commented on core/tools.js:
```diff
@@ -1488,6 +1488,42 @@
+		 * @param {String} handlerName Event handler attribute name (basically `on` + event name).
+		 * @param {Object} evt Event payload.
+		 */
+		fireElementEventHandler: function( element, handlerName, evt ) {
```
`element.fire` is connected with our own event system, so it's not a good place to add such functionality. OTOH `element.fireEventHandler` seems nice.

Comandeer replied:
Actually it probably makes more sense to broaden the scope of this function to just fire DOM event on the element with given payload 🤔. I'd prepare appropriate method.

Comandeer commented on plugins/menu/plugin.js:
```diff
@@ -108,11 +109,15 @@ CKEDITOR.plugins.add( 'menu', {
 
+	// We must block clicking with right mouse button (#2858).
+	if ( CKEDITOR.env.ie ) {
+		specialClickHandler = 'return false;" onmouseup="CKEDITOR.tools.getMouseButton(event)===CKEDITOR.MOUSE_BUTTON_LEFT&&';
```
Populating it during declaration would require introducing ternary operator there, to conditionally populate the variable only in IE. IMO the current code is more readable and consistent with other places, where such conditional is present (e.g. in `button` plugin).

Comandeer commented:
I've replaced most of magic numbers with bilateral mouse button normalization via newly added `CKEDITOR.tools.normalizeMouseButton` method. Additionally I've added `CKEDITOR.dom.element#fireEventHandler` method and fixed unit tests (as mouse event were dispatched from wrong `document` object, causing tests to always pass).

I didn't enable tests in any other browsers, as I don't see any reason to do it. The whole issue originates from the fact that we use `mouseup` event in IE, which is fired for every mouse click, not only ones made with primary button. In contrary, all other browsers use `click` event, which is [restricted only to primary button clicks](https://w3c.github.io/uievents/#event-type-click) (bold's mine):

> The `click` event type MUST be dispatched on the topmost event target indicated by the pointer, when **the user presses down and releases the primary pointer button**, or otherwise activates the pointer in a manner that simulates such an action.

There is no point of switching `click` to `mouseup` in any other browser, so IMO it's safe to ignore these tests in other browsers – as regression is very unlike to ever occur there.

Comandeer commented:
I see that introduced changes caused CI to fail due to keyboard handling in emoji. I'll fix it.

Comandeer commented:
I've fixed incorrect handling of keyboard in emoji plugin.

Comandeer commented:
> Tests checking if right click focuses item fails e.g.:
> `tests/plugins/listblock/manual/rightclick`
> `tests/plugins/menu/manual/rightclick`
> Right click doesn't focus clicked item, you can verify it by right-clicking item and pressing enter. In that case, the default option will be applied instead of the selected one. Also, there is no visible focus for context menu.

This is the mistake in test description, caused by the fact that visible focus indicator (outline) is in fact moved to the clicked item. I fixed tests descriptions.

I've also rebased the whole PR onto latest `major`.

jacekbogdanski commented on CHANGES.md:
```diff
@@ -12,6 +12,9 @@ Fixed Issues:
 * [#1478](https://github.com/ckeditor/ckeditor-dev/issues/1478): Fixed: Custom colors added to [Color Button](https://ckeditor.com/cke4/addon/colorbutton) via [`config.colorButton_colors`](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_config.html#cfg-colorButton_colors) in form label/code don't work correctly.
+* [#2845](https://github.com/ckeditor/ckeditor-dev/issues/2845): [Rich Combo](https://ckeditor.com/cke4/addon/richcombo) reacts to right mouse button click.
+* [#2857](https://github.com/ckeditor/ckeditor-dev/issues/2857): [List Block](https://ckeditor.com/cke4/addon/listblock) items react to right mouse button click.
+* [#2858](https://github.com/ckeditor/ckeditor-dev/issues/2858): [Menu](https://ckeditor.com/cke4/addon/menu) items react to right mouse button click.
```
I'm wondering about some entry change log merge here as they have mostly the same message. WDYT?

Comandeer replied:
I was also thinking about it, but I looked through the whole changelog and didn't spot anything like that. Because of that I decided to create separate entries.

Comandeer replied:
Actually I got an idea, how to do it. bed3e35058d2872fef3f5c3a335ac45f0338fe63 shows the proposed solution.

Comandeer commented:
> When right-clicking any toolbar button on Edge using manual tests, a selection from editor disappears.

It's reproducible also on `major`, I've created separated issue for it: #3170.

Comandeer commented on tests/plugins/menu/manual/rightclick.md:
```diff
@@ -0,0 +1,30 @@
+
+	Focus remains on the first option.
+
+	WARNING: it's best to check it using arrow keys, as visible focus indicator (background + outline) can be in fact moved to the clicked option.
```
The issue here is the fact that both Safari and Chrome applies `:active` pseudoclass when right-clicking: https://github.com/ckeditor/ckeditor-dev/issues/2845#issuecomment-466795603

Comandeer commented on core/dom/element.js:
```diff
@@ -2113,6 +2113,41 @@ CKEDITOR.dom.element.clearMarkers = function( database, element, removeFromDatab
+		 */
+		fireEventHandler: function( eventName, evt ) {
+			var handlerName = 'on' + eventName,
+				element = this.$;
```
I'm not a big fan of it. IMO the name of the variable is readable enough on its own. TBH I'm more eager to change to something like `nativeElement` than add `$` to the beginning – as it would be readable also for users that do not know the convention.

Comandeer commented:
> Looks good, just requires updating target version to 4.12.1 as we are already during release, so it won't make to 4.12.0.

Actually it seems to be targeted to 4.13.0, via milestone. I would also target it to new major release, as it introduces new feature, `CKEDITOR.dom.element#fireEventHandler`.

Comandeer commented:
I've rebased onto latest `major` and updated version tags. Additionally I removed all commits changing changelog and moved changelog entries into PR description.

f1ames commented:
Rebased on the latest `major`.

f1ames commented on plugins/richcombo/plugin.js:
```diff
@@ -34,11 +35,15 @@ CKEDITOR.plugins.add( 'richcombo', {
 		' onfocus="return CKEDITOR.tools.callFunction({focusFn},event);" ' +
-			( CKEDITOR.env.ie ? 'onclick="return false;" onmouseup' : 'onclick' ) + // https://dev.ckeditor.com/ticket/188
-				'="CKEDITOR.tools.callFunction({clickFn},this);return false;">' +
+		'onclick="' + specialClickHandler + 'CKEDITOR.tools.callFunction({clickFn},this);return false;">' +
```
Same as in menu plugin (see my comment in https://github.com/ckeditor/ckeditor-dev/pull/2859/files#diff-192937538423fc12567331ed28e3af35L115).

f1ames commented on tests/plugins/menu/menu.js:
```diff
@@ -38,10 +38,30 @@
+
+		// (#2858)
+		'test right-clicking menu items': function() {
+			if ( !CKEDITOR.env.ie ) {
```
@Comandeer can you refer to this comment? Is this similar situation as with other tests that you find the regressions very unlikely to happen?

Comandeer replied:
Yes, it's de facto impossible. We use `click` event in every other browser, so right click will never occur in the first place. In IE it occurred as we use `mouseup` event there due to some nasty bug in IE's native image controls.

f1ames commented on tests/plugins/richcombo/manual/rightclick.html:
```diff
@@ -0,0 +1,12 @@
+</div>
+
+<script>
+	if ( !CKEDITOR.env.ie ) {
```
I assume it's the same situation as in https://github.com/ckeditor/ckeditor-dev/pull/2859/files#r268396718.

f1ames commented on tests/plugins/richcombo/manual/rightclick.md:
```diff
@@ -0,0 +1,15 @@
+
+### Unexpected result
+
+Rich combo becomes open.
```
```suggestion
Rich combo opens.
```

Comandeer commented on tests/plugins/panel/manual/keydown.md:
```diff
@@ -0,0 +1,23 @@
+	### Unexpected
+
+	* Nothing happens.
+2. Repeat the same procedure for <kbd>Enter</kbd>.
```
Strangely it happens only in Firefox on macOS 🤔 I've filled a bug: https://github.com/ckeditor/ckeditor-dev/issues/3271

Comandeer commented on core/tools.js:
```diff
@@ -1462,23 +1462,73 @@
+		 * @param {Number} button Mouse button identifier.
+		 * @param {Boolean} [reverse=false] If set to true, the conversion is reversed: values
+		 * returned by other browsers are converted to IE 8 values.
+		 * @returns {Number} Normalized mouse button identifier.
```
TBH I think that fifth mouse button is unpopular enough to not need to worry about it. I'd leave it as is.

Comandeer commented on core/tools.js:
```diff
@@ -1462,23 +1462,73 @@
+		 * @returns {Number} Normalized mouse button identifier.
+		 */
+		normalizeMouseButton: function( button, reverse ) {
+			var mappings = [
```
Seems like supermicrooptimization 🤔 But yes, I can move it somewhere else.

Comandeer commented on tests/plugins/richcombo/richcombo.js:
```diff
@@ -56,5 +56,24 @@ bender.test( {
+
+	// (#2565)
+	'test right-clicking combo': function() {
+		if ( !CKEDITOR.env.ie ) {
```
As I stated there, I do not think such regression is even possible.

f1ames commented:
Rebased onto latest `major`.
